Key Completion Review

AdamRoden opened the pull request:
@JuliusSweetland Here you can take a look at how I propose implementing this change. It's actually a nice step toward unity between built-in and dynamic keyboard capability. I suggest copying the current field into the new setting on application launch.

JuliusSweetland commented:
I'll take a look as soon as possible, but I've got a full on week so please
bear with me.

On Mon, 19 Oct 2020, 19:25 AdamRoden, <notifications@github.com> wrote:

> @JuliusSweetland <https://github.com/JuliusSweetland> Here you can take a
> look at how I propose implementing this change. It's actually a nice step
> toward unity between built-in and dynamic keyboard capability. I suggest
> copying the current field into the new setting on application launch.
>
> TODO: Tweak MC name and description, add some input validation,
> commenting, other...?
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/OptiKey/OptiKey/pull/741
> Commit Summary
>
>    - Key Completion Review
>
> File Changes
>
>    - *M* src/JuliusSweetland.OptiKey.Core/Models/KeyValueAndTimeSpan.cs
>    <https://github.com/OptiKey/OptiKey/pull/741/files#diff-d52882ffa8f9224ebc6b44720f4c357cce4381b8a990b12842b630b5ce842a2e>
>    (10)
>    - *M*
>    src/JuliusSweetland.OptiKey.Core/Models/KeyValueAndTimeSpanGroup.cs
>    <https://github.com/OptiKey/OptiKey/pull/741/files#diff-64c07266a46296f16b0192256883cabfb2c95b45f8af611af23795c2ded40a66>
>    (8)
>    - *M*
>    src/JuliusSweetland.OptiKey.Core/Models/SerializableDictionaryOfTimeSpanByKeyValues.cs
>    <https://github.com/OptiKey/OptiKey/pull/741/files#diff-41f080cb42dbc20c24ac721f19cde8d3a093ac70c84026c74d9d7750590cf40a>
>    (13)
>    - *M*
>    src/JuliusSweetland.OptiKey.Core/Observables/TriggerSources/KeyFixationSource.cs
>    <https://github.com/OptiKey/OptiKey/pull/741/files#diff-e9b95dd8ad2edc339bf25f5079525deb21b26f3bce51c5ee7e010ec07a421768>
>    (76)
>    - *M* src/JuliusSweetland.OptiKey.Core/OptiKeyApp.cs
>    <https://github.com/OptiKey/OptiKey/pull/741/files#diff-79cffd90a8ef4da2a5b014f8cdd715f14d30a8d1485c55e9c5db4b4c7caeef17>
>    (9)
>    - *M* src/JuliusSweetland.OptiKey.Core/Properties/Resources.Designer.cs
>    <https://github.com/OptiKey/OptiKey/pull/741/files#diff-f199491a2ab93b0f1ffbbb1ffbd647a89adbea48c464d4b41903cf73478e0f1e>
>    (2)
>    - *M* src/JuliusSweetland.OptiKey.Core/Properties/Resources.resx
>    <https://github.com/OptiKey/OptiKey/pull/741/files#diff-9180e75b9604d425fae02784e0e3b2d419a3a09e346de5f56f7f3da3be277042>
>    (2)
>    - *M* src/JuliusSweetland.OptiKey.Core/Properties/Settings.cs
>    <https://github.com/OptiKey/OptiKey/pull/741/files#diff-5985dc60b992248bdbd803b440f58b8d4df32e23a03372107461fefb0c6a7aa5>
>    (24)
>    - *M*
>    src/JuliusSweetland.OptiKey.Core/UI/ViewModels/Management/PointingAndSelectingViewModel.cs
>    <https://github.com/OptiKey/OptiKey/pull/741/files#diff-aca24e103f6f90daedc71795b9aa6e92b1a692302466b7b7d8a5f3896ec969ff>
>    (363)
>    - *M*
>    src/JuliusSweetland.OptiKey.Core/UI/Views/Management/PointingAndSelectingView.xaml
>    <https://github.com/OptiKey/OptiKey/pull/741/files#diff-b067150ec3194dacb45e59ae5190ec7f9b6f33026f8b460c1b54d1a9ab856fe0>
>    (27)
>
> Patch Links:
>
>    - https://github.com/OptiKey/OptiKey/pull/741.patch
>    - https://github.com/OptiKey/OptiKey/pull/741.diff
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/OptiKey/OptiKey/pull/741>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABEA7APARMM3JLDO4UPWAPLSLSAANANCNFSM4SWQB3KQ>
> .
>

JuliusSweetland commented on src/JuliusSweetland.OptiKey.Core/OptiKeyApp.cs:
```diff
@@ -115,6 +115,18 @@ protected void Initialise()
+            if (timespan> TimeSpan.Zero)
+            {
+                Settings.Default.KeySelectionTriggerFixationDefaultCompleteTimes = timespan.TotalMilliseconds.ToString()
+                    + "," + Math.Round(0.75 * timespan.TotalMilliseconds).ToString()
```
I'm aiming to look at this PR (and your other one) in the next few days. Can you clarify what this logic is doing?

AdamRoden replied:
I can clarify the comment before finalizing the PR, but the intent here was to convert the old single value Completion Time into a list with the new format to nudge users toward exploring the feature. Another idea is to add conditional highlighting to the field in MC and show a message that repeat rates can be added.

JuliusSweetland commented:
Ah I see. If this is for all key selection times then I wouldn't default
their old value to one which increases in selection speed. I fear some
users will not understand why keys are suddenly being selected multiple
times in faster succession. I think a mask on the textbox + a hint would go
a long way though.

On Thu, Oct 29, 2020 at 11:57 AM AdamRoden <notifications@github.com> wrote:

> *@AdamRoden* commented on this pull request.
> ------------------------------
>
> In src/JuliusSweetland.OptiKey.Core/OptiKeyApp.cs
> <https://github.com/OptiKey/OptiKey/pull/741#discussion_r514202586>:
>
> > @@ -115,6 +115,18 @@ protected void Initialise()
>                  Settings.Default.Reload();
>              }
>
> +            //Migrate the completion time setting to the new field
> +            var timespan = Settings.Default.KeySelectionTriggerFixationDefaultCompleteTime;
> +            if (timespan> TimeSpan.Zero)
> +            {
> +                Settings.Default.KeySelectionTriggerFixationDefaultCompleteTimes = timespan.TotalMilliseconds.ToString()
> +                    + "," + Math.Round(0.75 * timespan.TotalMilliseconds).ToString()
>
> I can clarify the comment before finalizing the PR, but the intent here
> was to convert the old single value Completion Time into a list with the
> new format to nudge users toward exploring the feature. Another idea is to
> add conditional highlighting to the field in MC and show a message that
> repeat rates can be added.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/OptiKey/OptiKey/pull/741#discussion_r514202586>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABEA7AKK23KQP3FQ3DWIM7TSNFKCRANCNFSM4SWQB3KQ>
> .
>

AdamRoden commented:
I added additional language to help users understand what they can enter, highlighted the field if it does not have multiple values, and added input validation.

JuliusSweetland commented:
@AdamRoden Very nice PR. Merged. Thank you.
